### ISSUE TITLE
Not all pokemon are being sent to all webhooks [Bug - User Specific]

### ISSUE THREAD

**Issue by TackyBrad** [kvangent/PokeAlarm#281]
## Problem Description

I had been running three webhooks properly from each of my scans all outputting to different twitters. The other day I noticed one twitter did not pop when it was supposed to so I investigated the logs and found it just absent from the webhook, but the same pokemon was present in another webhook. Further, around the same time, the one that had the absent pokemon had found a spearow the other webhook (that did find the first poke) did not find.

I hope that mess of modifiers is understandable. Basically, WH 1 Did not find poke 43, but WH 2 did. Seconds later, WH 1 found poke 57, but WH 2 did not.
## Your Environment

Operating System:  Windows
Python Version: 2.7 
PokemonGo-Map startup command: `Python runserver.py -ng –ac acc.csv  –l “lat/long” -sn statusname –st 10 –k redacted –sd 30 –j –nk –-wh-threads 500 -wh http://127.0.0.1:4000 http://127.0.0.1:4001 http://127.0.0.1:4002 http://127.0.0.1:4003 -w 50 –-db-threads 200 -ss -ns
`  
PokeAlarm startup command: `python runwebhook.py` `python runwebhook.py -P 4001` `python runwebhook.py -P 4002`

Did the issue persist after restarting the script?: Yes

## Console Output

https://s16.postimg.org/5evy46qt1/Pokealarm_Sideby_Side.png

## Your Alarms Config

```json
    {
	"alarms":[
		{
			"active":"False",
			"type":"pushbullet",
			"api_key":"YOUR_API_KEY"
		},
		{
			"active": "False",
			"type":"slack",
			"api_key":"YOUR_API_KEY",
			"channel":"#general"
		},
		{
			"active": "False",
			"type": "pushover",
			"app_token": "YOUR_APP_TOKEN",
			"user_key": "YOUR_USER_KEY"
		},
		{
            "active": "True",
            "type": "twitter",
            "access_token": "hghg",
            "access_secret": "ghg",
            "consumer_key": "ghg",
            "consumer_secret": "ghg"
        }
	],
	"gyms":{
		"To_Valor":"False",
		"To_Mystic":"False",
		"To_Instinct":"False",
		"From_Valor":"False",
		"From_Mystic":"False",
		"From_Instinct":"False"
	},
	"pokestops":{
		"Lured":"false"
	},
	"pokemon":{			
		"Bulbasaur":{ "min_iv":"98" },
		"Ivysaur":{ "min_iv":"98" },
		"Venusaur":{ "min_iv":"98" },
		"Charmander":{ "min_iv":"98" },
		"Charmeleon":{ "min_iv":"98" },
		"Charizard":{ "min_iv":"98" },
		"Squirtle":{ "min_iv":"98" },
		"Wartortle":{ "min_iv":"98" },
		"Blastoise":{ "min_iv":"98" },
		"Caterpie":"False",
		"Metapod":"False",
		"Butterfree":"False",
		"Weedle":"False",
		"Kakuna":"False",
		"Beedrill":"False",
		"Pidgey":"False",
		"Pidgeotto":"False",
		"Pidgeot":"False",
		"Rattata":"False",
		"Raticate":"False",
		"Spearow":"False",
		"Fearow":"False",
		"Ekans":"False",
		"Arbok":"False",
		"Pikachu":"False",
		"Raichu":"False",
		"Sandshrew":"False",
		"Sandslash":"False",
		"Nidoran?":"False",
		"Nidorina":"False",
		"Nidoqueen":"False",
		"Nidoran?":"False",
		"Nidorino":"False",
		"Nidoking":"False",
		"Clefairy":"False",
		"Clefable":"False",
		"Vulpix":"False",
		"Ninetales":"False",
		"Jigglypuff":"False",
		"Wigglytuff":"False",
		"Zubat":"False",
		"Golbat":"False",
		"Oddish":"False",
		"Gloom":"False",
		"Vileplume":"False",
		"Paras":"False",
		"Parasect":"False",
		"Venonat":"False",
		"Venomoth":"False",
		"Diglett":"False",
		"Dugtrio":"False",
		"Meowth":"False",
		"Persian":"False",
		"Psyduck":"False",
		"Golduck":"False",
		"Mankey":"False",
		"Primeape":"False",
		"Growlithe":{ "min_iv":"98" },
		"Arcanine":{ "min_iv":"95" },
		"Poliwag":"False",
		"Poliwhirl":"False",
		"Poliwrath":"False",
		"Abra":{ "min_iv":"95" },
		"Kadabra":"False",
		"Alakazam":"False",
		"Machop":{ "min_iv":"95" },
		"Machoke":"False",
		"Machamp":"True",
		"Bellsprout":"False",
		"Weepinbell":"False",
		"Victreebel":"False",
		"Tentacool":"False",
		"Tentacruel":"False",
		"Geodude":"False",
		"Graveler":"False",
		"Golem":"False",
		"Ponyta":"False",
		"Rapidash":"False",
		"Slowpoke":"False",
		"Slowbro":"False",
		"Magnemite":"False",
		"Magneton":"False",
		"Farfetch'd":"False",
		"Doduo":"False",
		"Dodrio":"False",
		"Seel":{ "min_iv":"98" },
		"Dewgong":"False",
		"Grimer":"False",
		"Muk":"False",
		"Shellder":"False",
		"Cloyster":"False",
		"Gastly":"False",
		"Haunter":"False",
		"Gengar":"False",
		"Onix":"False",
		"Drowzee":"False",
		"Hypno":"False",
		"Krabby":"False",
		"Kingler":"False",
		"Voltorb":"False",
		"Electrode":"False",
		"Exeggcute":{ "min_iv":"98" },
		"Exeggutor":{ "min_iv":"98" },
		"Cubone":"False",
		"Marowak":"False",
		"Hitmonlee":"False",
		"Hitmonchan":"False",
		"Lickitung":"False",
		"Koffing":"False",
		"Weezing":"False",
		"Rhyhorn":{ "min_iv":"95" },
		"Rhydon":{ "min_iv":"95" },
		"Chansey":"False",
		"Tangela":"False",
		"Kangaskhan":"False",
		"Horsea":"False",
		"Seadra":"False",
		"Goldeen":"False",
		"Seaking":"False",
		"Staryu":"False",
		"Starmie":"False",
		"Mr. Mime":"False",
		"Scyther":{ "min_iv":"98" },
		"Jynx":"False",
		"Electabuzz":"False",
		"Magmar":"False",
		"Pinsir":"False",
		"Tauros":"False",
		"Magikarp":"False",
		"Gyarados":{ "min_iv":"95" },
		"Lapras":"True",
		"Ditto":{ "min_iv":"95" },
		"Eevee":{ "min_iv":"98" },
		"Vaporeon":{ "min_iv":"95" },
		"Jolteon":"False",
		"Flareon":"False",
		"Porygon":"True",
		"Omanyte":"False",
		"Omastar":"False",
		"Kabuto":"False",
		"Kabutops":"True",
		"Aerodactyl":"True",
		"Snorlax":{ "min_iv":"95" },
		"Articuno":"True",
		"Zapdos":"True",
		"Moltres":"True",
		"Dratini":{ "min_iv":"97" },
		"Dragonair":"False",
		"Dragonite":"True",
		"Mewtwo":"True",
		"Mew":"True"
	}
}

```

## Troubleshooting Checklist
[x ] Most recent version of PokemonGo-Map
[ x] Most recent version of PokeAlarm
[ ] Checked the Wiki page for service in question

**Comment by kvangent** [comment]
How does PokemonGo-Map handle time now? What does it send for expire time?

**Comment by TackyBrad** [comment]
I'm using a PR that uses algorithms to determine accurate spawn timers after scanning an area for a while. Before that is effective, it defaults to 30 minutes.

**Comment by Albert62** [comment]
@kvangent +14min to 15min

**Comment by kvangent** [comment]
Do you have the debug logs for PokeAlarm for this incident?

**Comment by filippounitn** [comment]
can say that's true. Using last version of pokemongomap with -speed. Lost a snorlax. It showed up on the map but didn't trigger the the telegram notification. I'm looking for th elog file but i can't find it. Do i need to put something in the command to create one ? Otherwise just tell me where to find that file please. 
(i'm using a old version of pokealarm but before today there has never been a problem)

**Comment by kvangent** [comment]
The log file is whatever gets outputted to the console. You can append `--debug` to get some more verbose logging.

I am extremely skeptical that the issue is on PokeAlarm's side without more information .

**Comment by filippounitn** [comment]
Yeah, problably the problem is not pokealarm, this was a new pr that was recently merged and probably caused some problems. 
I have no log file, I tried to look up the console but it was too late and it got cancelled or I coudn't find it. It would be awesome if there is a way to get a txt file with the log so it can be stored.

**Comment by TackyBrad** [comment]
@kvangent What would you like me to do, should I add --debug to the code string and monitor to see what is going on? Am I running too many webhooks from the map (4) that it is losing data (even though only two are being listened on)? Just let me know what course of action you think I should take.

Thanks for such an amazing service.

**Comment by kvangent** [comment]
@TackyBrad Try in debug mode and see if you can tell if there are any behind the screen errors being thrown or web hooks that don't seem to be getting processed.

**Comment by Imoq** [comment]
So, I need to piggyback on this report (perhaps is no longer "user specific" anymore).

I am using latest develop version for the map, no PRs. 99%> of my spawnpoints have TTH identified.

Today I had two cases where important notifications were lost.

First case: a -st 5 area that I run with two webhooks (one for Twitter/Facebook (brand new, yay!) and the second for Telegram). On the Telegram I received the notification for a Dragonite that was there when I went to catch it. But the notification was never triggered on Twitter or Facebook page. I imagined it was just a temporary mistake and ignored it.

Later tonight, a Snorlax appeared on the map, but it was with an -st 7 area with only one webhook (Twitter) configured. The Twitter notification never fired off and I was actually able to look at the time when (acording to the DB, I did the queries) the Pokémon was identified and inserted. The regular log (no debug or verbose) for PokéAlarm didn't identify it so, is it the map that didn't send it?

I have a copy of the regular PokeAlarm log for that period of time, as well as the DB queries where the Snorlax appeared, if necessary. I also have a log of all my running instances and I grepped for "webhooks" in case I did see any error message, but there wasn't any (like, related to webhook queue, for example).

The only configuration I have relevant that may or may not be culprit here is this on config.ini:

> webhook-updates-only: TRUE

Please, @kvangent let me know if I can provide any more information to debug but seeing that this is an issue I have detected with two "big" Pokémon perhaps this is going on as well on regular ones and we just hadn't noticed until now.

Thanks!

**Comment by kvangent** [comment]
@Imoq So you had one where Telegram fired off, but Twitter and Facebook didnt? Was this all on the same instance of PokeAlarm?  Do you know what kind of spawn it was? Is it possible it was one of the spawns/dispawns where it spawned earlier and then spawned a 2nd time?

**Comment by Imoq** [comment]
Correct, Telegram did fire off but Twitter/Facebook did not. They are on two different instances of PokeAlarm (two different webhooks, one for Twitter/Facebook, the second for Telegram), the parameter I use to run that instance is:

`-wh http://127.0.0.1:4000 http://127.0.0.1:4100`

The first one is Twitter/FB, the second one is Telegram.

For the kind of spawn, I don't know if this info is useful:

```
+------------------------------+---------------+------------+------------------+------------------+---------------------+-------------------+--------------------+--------------------+--------+--------+---------------------+
| encounter_id                 | spawnpoint_id | pokemon_id | latitude         | longitude        | disappear_time      | individual_attack | individual_defense | individual_stamina | move_1 | move_2 | last_modified       |
+------------------------------+---------------+------------+------------------+------------------+---------------------+-------------------+--------------------+--------------------+--------+--------+---------------------+
| MTc3Mzk4MjkyNDc0ODM4Nzg0ODc= | 12b5f202387   |        143 | 43.9167651498582 | 4.90000921970679 | 2016-12-19 00:17:39 |                 8 |                  4 |                  1 |    234 |     14 | 2016-12-18 23:18:09 |
```

For what is worth, on the database it doesn't exist an earlier or a later pop recorded for that pokemon ID, I'll go ahead and paste the complete query, it's not so long but apologies for the lenght in advance.

```
mysql> select * from pokemon where pokemon_id = 143;
+------------------------------+---------------+------------+------------------+------------------+---------------------+-------------------+--------------------+--------------------+--------+--------+---------------------+
| encounter_id                 | spawnpoint_id | pokemon_id | latitude         | longitude        | disappear_time      | individual_attack | individual_defense | individual_stamina | move_1 | move_2 | last_modified       |
+------------------------------+---------------+------------+------------------+------------------+---------------------+-------------------+--------------------+--------------------+--------+--------+---------------------+
| MTc3MjgyMjQ3MTk0NDU3OTI4MjI= | 12c9c7772f5   |        143 | 43.2460203804676 | 5.38123629874774 | 2016-12-17 10:17:10 |                 3 |                 12 |                 11 |    212 |    131 | 2016-12-17 09:47:43 |
| MTc3Mzk4MjkyNDc0ODM4Nzg0ODc= | 12b5f202387   |        143 | 43.9167651498582 | 4.90000921970679 | 2016-12-19 00:17:39 |                 8 |                  4 |                  1 |    234 |     14 | 2016-12-18 23:18:09 |
| MTEwNjcyNjI0ODYzODYwMTM3NQ== | 12c9bdcfda9   |        143 | 43.3216334846265 |  5.4888004434144 | 2016-12-16 13:11:29 |                 7 |                 14 |                  8 |    212 |     14 | 2016-12-16 12:47:20 |
| MTIwMDQ2MjA2ODMwMTY1ODk4MTk= | 12c9c0b414b   |        143 | 43.2841760164832 | 5.37722901384141 | 2016-12-17 18:33:41 |                 2 |                 10 |                 14 |    234 |    131 | 2016-12-17 18:05:10 |
| MTM1MDg3NjMzMTM0NDYyMzM4NTc= | 12b5edff6b3   |        143 | 43.9188055296911 | 4.89230594515729 | 2016-12-18 02:56:50 |                 5 |                 10 |                 12 |    212 |     14 | 2016-12-18 02:28:18 |
| MTY1NzE0MjU1MDIwMTkzNzI0NA== | 12c9e442c27   |        143 | 43.4719166246579 |  5.1658122710943 | 2016-12-16 23:53:36 |                 2 |                  0 |                  9 |    234 |     14 | 2016-12-16 23:52:26 |
| MTY5ODE0NTIyOTEwMjgzNTYzNTc= | 12c9bdc551d   |        143 | 43.3239697508334 | 5.48871850759911 | 2016-12-17 14:35:42 |                 7 |                  8 |                 13 |    234 |     14 | 2016-12-17 14:21:27 |
| NjU0MDEwMzg4MjAwMjYwMzk1     | 12c9c768f2f   |        143 | 43.2581134967573 | 5.38205414614048 | 2016-12-17 02:53:57 |                 4 |                  2 |                 14 |    234 |    131 | 2016-12-17 02:24:10 |
| NTA3MzA3Nzg3OTAxMzU2MzM0MQ== | 12c9bf12c49   |        143 | 43.3059198593573 | 5.41584134826575 | 2016-12-18 04:21:06 |                11 |                  8 |                  8 |    212 |    131 | 2016-12-18 03:51:20 |
| NTg3Njg0Mzg2ODQ2MzY1NDUzNg== | 12c9c0c8899   |        143 | 43.2896227232672 | 5.37027826056989 | 2016-12-19 00:47:50 |                12 |                  4 |                  0 |    234 |     14 | 2016-12-19 00:18:33 |
| NTgyOTc5MTM4MjY2OTAxNjQxMg== | 12c9c7434ab   |        143 | 43.2646024611336 | 5.38344451327047 | 2016-12-16 20:04:49 |                12 |                  1 |                  1 |    234 |     31 | 2016-12-16 19:36:46 |
| Nzk0NzAxNzY3NjAwOTI2MjA2Nw== | 12c9c0ca681   |        143 | 43.2838308462287 | 5.37175011497148 | 2016-12-16 14:57:39 |                 1 |                  3 |                 11 |    234 |     14 | 2016-12-16 14:56:42 |
| ODA3NDI0ODM5NzgwODgyNTcyNQ== | 12c9c76aca5   |        143 | 43.2601105949229 | 5.37322200689092 | 2016-12-17 21:02:19 |                 4 |                 11 |                 12 |    212 |     14 | 2016-12-17 20:32:42 |
| ODQyMjkzNjYxNzIwNDQyMDg4MA== | 12c9bf62433   |        143 | 43.3055272493129 | 5.39546849445508 | 2016-12-16 13:32:09 |                 2 |                 12 |                 14 |    234 |    131 | 2016-12-16 12:42:54 |
| ODY3MzA2MDAxODI1OTQxODgwNA== | 12c9bf1296b   |        143 | 43.3067168170756 | 5.41649601661841 | 2016-12-18 15:56:15 |                 1 |                  1 |                 10 |    212 |    131 | 2016-12-18 15:41:28 |
| OTg5MTExNjk2ODM1MDEzMTM4Mg== | 12c9c7514db   |        143 | 43.2744573693052 | 5.38851554739544 | 2016-12-18 15:41:34 |                14 |                  5 |                  7 |    234 |    131 | 2016-12-18 15:12:16 |
+------------------------------+---------------+------------+------------------+------------------+---------------------+-------------------+--------------------+--------------------+--------+--------+---------------------+
16 rows in set (0.00 sec)

mysql> 
```

I don't know if I can provide more information, but I'll gladly try. I also can look for the DB info for the Dragonite that was notified via Telegram but not via Twitter/FB.

Thanks!

**Comment by kvangent** [comment]
@Imoq Can you look back at your chat history and see if there was a Snorlax posted with a similar location?

**Comment by Imoq** [comment]
Sorry, my computer crashed and I just went to bed, replying from phone.
There isn't any other Snorlax reported on that location via any way
(Twitter, FB or Telegram) and it didn't appear on the console output for
PokeAlarm, like if it never received the data from the map (for the Twitter
/ FB webhook, that is).

I hope that is what you were asking, apologies but I'll stop answering for
few hours.

Thank you!

**Comment by kvangent** [comment]
@Imoq And there were no errors reported  via the console output?

There is a possibility that the map sent to one webhook but not the other one. It is impossible to know or sure because the map doesn't wait to make sure the webhook landed. What OS are you running on?

**Comment by Imoq** [comment]
Zero errors in console output both for the map and for the PokeAlarm.

Running Linux, Ubuntu 16.04.

I have the feeling this may be more or less constant but we hadn't realized
until the "big" Pokémon were missing. I may for sure add any require
verbose tomorrow to try to figure out if there is something hidden.

Thanks!

**Comment by Imoq** [comment]
Today I detected two new cases of 2 Dragonites, under two different webhooks tied to the same process of runserver.py. The first one (Telegram) didn't notify, the second one (Twitter+FB+Telegram) did notify on all 3. Definitely something is going on, and these ones we detect because these are big or important Pokémon but I believe that if we run tests with smaller ones we can get some data.

Any ideas on what kind of data I can get to share and see if we can get to the core of this?

**Comment by sclo012** [comment]
I have also noticed this problem from before, back when we had the old .35 api where I ran multiple webhooks. However, I can't reproduce this problem when I run only 1 webhook for the current (.45) api, so maybe that has something to do with using multiple webhooks? How would you detect the problem if you only run 1 webhook anyway?

**Comment by kvangent** [comment]
So essentially the problem is that PokemonGo-Map doesn't check to make sure that the target server actually received the webhook, or even check to make sure it was accepted correctly. Also, due to the way the error catching is set up, if anything bad happens Pokemon-Go Map just ignores it.

What needs to happen is PokemonGo-Map needs to have the webhook portion rewritten to be more robust.

**Comment by Buncertainc** [comment]
I have also find this problem.
After a notification is triggered, some pokemons sometimes are missing in the cmd window.
Please find the below screenshots of two webhooks for reference.
![default](https://cloud.githubusercontent.com/assets/23481910/21470237/7cf60dc6-caba-11e6-9edc-e3d5554dffd3.png)
![2](https://cloud.githubusercontent.com/assets/23481910/21470238/7cf708a2-caba-11e6-8595-9ba2b6fcade9.png)

**Comment by kvangent** [comment]
Essentially we need to make some modifications to PokeAlarm to see if we can diagnose what is happening in these instances. I'll see if I can add a few tweaks to get some better logging and see where the communication is breaking down.

**Comment by Imoq** [comment]
Thank you for your hard work, @kvangent ! It's really appreciated.

**Comment by sebastienvercammen** [comment]
As a side note, I've made some changes to webhooks in PokemonGo-Map: increased base timeout, proper logging of exceptions with trace, and all requests will now auto-retry up to 5 times (backoff factor 0.25s) in case of errors (both exceptions/timeouts and common 50x HTTP error codes).

Specifics are here, currently being tested, but to be merged soon: https://github.com/PokemonGoMap/PokemonGo-Map/pull/1666

Hope that helps @kvangent

**Comment by kvangent** [comment]
@sebastienvercammen Great, these were more or less the same as some changes I was just about to sit down and do.

@Imoq @TackyBrad You guys should give that branch a try and see if that addresses the problem.

**Comment by catnet23** [comment]
@TackyBrad & @Imoq 
looks like you got the wrong syntax. It's not just one -wh you need for every webhook a -wh

commandline:
-wh http://127.0.0.1:1345 -wh http://127.0.0.1:12346

**Comment by TackyBrad** [comment]
That literally just changed in one of the most recent commits from a PR
that was started 11 days ago. Our problem vastly predates that, though it's
possible their fix, which now caused you to need to do a -wh for each,
solved the problem.

**Comment by catnet23** [comment]
@TackyBrad the syntax changed because it never works with the config file.
after the fix that it work now with the config file, the syntax changed to the normal syntax like proxy.

**Comment by Buncertainc** [comment]
I am not sure if my problem is the same as @TackyBrad 's
My problem is that when the pokealarm is processing a trigger and during the processing time, most likely a few seconds, there is new information comes from pokemongomap, the pokealarm cannot handle the new information. Only after the triggering processing is finished, the pokealarm starts handle another new information again.
Below is the console when the pokealarm is processing a trigger. The red square,I believe, is the information comes from pokemongomap during the trigger process
![4](https://cloud.githubusercontent.com/assets/23481910/21606626/bbb99bee-d1eb-11e6-9f8f-8f0ad505f780.png)

**This is the normal handling process, I believe, of pokealarm for a requset**
![6](https://cloud.githubusercontent.com/assets/23481910/21606727/52e2b654-d1ec-11e6-9b4b-7b06d868c424.png)

**This is the pokemongomap console after the update of webhook.py** 
![5](https://cloud.githubusercontent.com/assets/23481910/21606682/fda93a78-d1eb-11e6-8775-17caf51bade6.png)
I am not sure if this is only my problem or other users also have this problem but I hope that it can help to know what is happening.

**Comment by sebastienvercammen** [comment]
@TackyBrad, the syntax change was a fix for multiple webhooks not working via config file.

PR 1666 is the actual "fix" you're looking for (and even then it's technically not a fix because nothing was broken, rather it's an improvement).

And I'm sure @catnet was just clarifying for future use, not saying that it was the cause of your issue. Just a language barrier problem.

But all of this is not related to PokeAlarm, so let's let @kvangent focus on doing his thing. :) The PGM Discord is open if anyone has more questions.

**Comment by kvangent** [comment]
@TackyBrad @Imoq Are we still having issues or did this fix the problem?

**Comment by Imoq** [comment]
@kvangent I haven't seen the issue anymore since Seb's fix. I have compared both alerting systems for two different webhooks and I don't see any Pokémon that appears on one but doesn't on the other. Furthermore, I haven't had any complain from the people using the maps so I imagine it is fixed for good. Of course, no formal tests here, just the anecdotal data but I believe it should be good for now, unless I can run any specific test to save and compare the data.

Thank you both!

**Comment by kvangent** [comment]
I'm going to close this out. If any users continue to find issues, please let me know.
